### PR TITLE
Class procedure support for template

### PR DESCRIPTION
**Pull request by ansharlubis**
Work supporting the example described in #2265.

The ASR is built and verified, but class procedure calls seem to fail on LLVM.

### CODE REVIEW

**Comment by certik** [comment]
Very nice, thanks!

File: tests/reference/asr-template_array_04-ed63449.stdout [hunk 1]
```diff
@@ -371,14 +371,22 @@
...
                                         default_behavior
                                         [t]
+                                    )
+                                    (Require
+                                        default_behavior
+                                        [t]
+                                    )
+                                    (Require
+                                        default_behavior
+                                        [t]
```
Comment (certik) [inline tests/reference/asr-template_array_04-ed63449.stdout:381]: It looks like these `(Require` statements tripled. That might be a bug.

**Review by certik — APPROVED** [review]
Otherwise it looks good, thanks!

**Comment by certik** [comment]
Also, please open up an issue on the bug of calling "push_back".

**Comment by ansharlubis** [comment]
Moving the refactorization to a different PR because it got messy.

**Pull request closed without merging**